tests/BaseV3.php

```diff
@@ -0,0 +1,264 @@
+            CURLOPT_HTTPHEADER => $headersParsed
+        );
+        
+        \curl_setopt_array($ch, $optArray);
+
+        $body = curl_exec($ch);
+        $code = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
+
+        \curl_close($ch);
+
+        return [
+            'code' => $code,
+            'body' => $body,
+            'headers' => $responseHeaders
+        ];
+    }
+
+    public function testPlaintextResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'plaintextResponse']);
+        self::assertEquals(200, $response['code']);
+        self::assertEquals('Hello World 👋', $response['body']);
+    }
+
+    public function testJsonResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'jsonResponse']);
+        self::assertEquals(200, $response['code']);
+        self::assertEquals('application/json', $response['headers']['content-type']);
+
+        $body = \json_decode($response['body'], true);
+
+        self::assertEquals(true, $body['json']);
+        self::assertEquals('Developers are awesome.', $body['message']);
+    }
+
+    public function testRedirectResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'redirectResponse']);
+        self::assertEquals(301, $response['code']);
+        self::assertEmpty($response['body']);
+        self::assertEquals('https://github.com/', $response['headers']['location']);
+    }
+
+    public function testEmptyResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'emptyResponse']);
+        self::assertEquals(204, $response['code']);
+        self::assertEmpty($response['body']);
+        self::assertEquals('0', $response['headers']['content-length']);
+    }
+
+    public function testNoResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'noResponse']);
+        self::assertEquals(500, $response['code']);
+        self::assertEmpty($response['body']);
+        self::assertStringContainsString('Return statement missing. return context.res.empty() if no response is expected.', $response['headers']['x-open-runtimes-errors']);
+    }
+
+    public function testDoubleResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'doubleResponse']);
+        self::assertEquals(200, $response['code']);
+        self::assertEquals('This should be returned.', $response['body']);
+    }
+
+    public function testHeadersResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'headersResponse', 'x-open-runtimes-custom-in-header' => 'notMissing']);
+        self::assertEquals(200, $response['code']);
+        self::assertEquals('OK', $response['body']);
+        self::assertEquals('first-value', $response['headers']['first-header']);
+        self::assertEquals('missing', $response['headers']['second-header']);
+        self::assertArrayNotHasKey('x-open-runtimes-custom-out-header', $response['headers']);
```

Comment:
So custom headers cant be sent as response?

Reply:
They can, but they cannot start with `x-open-runtimes-`. Notice `first-header` has test for a proper value from: https://github.com/open-runtimes/open-runtimes/blob/d7549b672133696085ac26e4c3da71877cecca12/tests/node-18.0/tests.js#L22

Reply:
`x-open-runtimes` is reserved keyword here. Same happens when request comes in. `context.req.headers` will not include any of headers starting with `x-open-runtimes-`.

Code for filtering request headers:

https://github.com/open-runtimes/open-runtimes/blob/3471b94b41f881df4f12cb1745afc9d63ebdfe94/runtimes/node-18.0/server.js#L28-L31

Code for filtering response headers:

https://github.com/open-runtimes/open-runtimes/blob/3471b94b41f881df4f12cb1745afc9d63ebdfe94/runtimes/node-18.0/server.js#L124-L130

Reply:
Developers can sand almost any headers. You can do `a`, `bla`, `what-ever-you-want`.

What you cannot do is header like `x-open-runtimes-something-bla`. That will not be passed into context for security reasons.

Regarding `x-open-runtimes-secret`, yes, that needs to be provided otherwise 401 is shown. Tho again, such a secret should bot be visible in the context. So value of this header is removed from context in order for developer to not see this secret.
For example, in Appwrite secret is between runtime and executor. There is no need for developer to know it. Even worse, if developer knew the secret, it would be security vulnerability possibly.